### PR TITLE
Add logging

### LINKED ISSUE #16: Add detailed logging to workflows instead of printing
Use the `logging` library to setup a logging mechanism to show the process of functions being run.
* [Logging documentation](https://docs.python.org/3/library/logging.html)
* [Gist of logging script](https://gist.github.com/nguyenkims/e92df0f8bd49973f0c94bddf36ed7fd0)

### PR DESCRIPTION
**Pull request by sometheasiekswx**
**Summary of change:**

Major:

- Added Logger
- Replaced all printing with logging

---

**This Change:**

- [x] is ready for testing
- [ ] a partial commit and not ready for testing
- [ ] introduces diagnostic functionality to further investigate an issue
- [ ] for review only and should not be merged
- [ ] other: please specify

---

**Related Work Items:**

#16 

---

**The following developer testing has been completed:**

Tested by running the program on local host.

---

**Deployment Notes:**

### CODE REVIEW

**Comment by aidanjgriffiths** [comment]
@sometheasiekswx note the CI is failing because you are writing to a log file by default which causes this error:
```python
FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/speaker-verification/speaker-verification/speaker_verification/logs/logs-23-04-2021.log'
```
I recommend by default the logging should be output to the command line and creating a command line argument to write to a log file.

**Comment by sometheasiekswx** [comment]
> ```python
> FileNotFoundError
> ```

I don't think logging should be done via the command line as we want it to be run within our code. Logging does output to the command line as well as write to the log.

**Comment by sometheasiekswx** [comment]
> @sometheasiekswx note the CI is failing because you are writing to a log file by default which causes this error:
> 
> ```python
> FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/speaker-verification/speaker-verification/speaker_verification/logs/logs-23-04-2021.log'
> ```
> 
> I recommend by default the logging should be output to the command line and creating a command line argument to write to a log file.

I don't think logging should be done via the command line as we want it to be run within our code. Logging does output to the command line as well as write to the log.

**Comment by aidanjgriffiths** [comment]
> I don't think logging should be done via the command line as we want it to be run within our code. 

I’m not sure what you mean here.
>Logging does output to the command line as well as write to the log.

Ok thats good, but outputting to a logfile shouldn't be the default behaviour, otherwise unit tests will log seemly random events or in the case of the CI outputting to a file that doesn’t exist. Regardless the CI needs to be passing in order to merge the PR so theres still work to be done. 

You may also want to look at squashing your commits into 1 commit since they all have the same commit message.

**Comment by sometheasiekswx** [comment]
> > I don't think logging should be done via the command line as we want it to be run within our code.
> 
> I’m not sure what you mean here.
> 
> > Logging does output to the command line as well as write to the log.
> 
> Ok thats good, but outputting to a logfile shouldn't be the default behaviour, otherwise unit tests will log seemly random events or in the case of the CI outputting to a file that doesn’t exist. Regardless the CI needs to be passing in order to merge the PR so theres still work to be done.
> 
> You may also want to look at squashing your commits into 1 commit since they all have the same commit message.

Squashing the commits is a good idea, I'll do that. Still not sure why the file path keeping having issues when I works on my local machine...

**Comment by aidanjgriffiths** [comment]
> Still not sure why the file path keeping having issues when I works on my local machine...

Try removing the local log file, and then run the unit tests to see if that fails. I believe the logger doesn't execute the code that creates the logfile in the case of the unit tests so its logging to a file that doesn't exist.

**Comment by sometheasiekswx** [comment]
> > Still not sure why the file path keeping having issues when I works on my local machine...
> 
> Try removing the local log file, and then run the unit tests to see if that fails. I believe the logger doesn't execute the code that creates the logfile in the case of the unit tests so its logging to a file that doesn't exist.

There is no log file committed since it's ignored. I think it might be with the logs folder not existing. I've added a .gitkeep to create the logs folder. Hopefully, that works :pray: :pray: :pray:

**Comment by sometheasiekswx** [comment]
> > > Still not sure why the file path keeping having issues when I works on my local machine...
> > 
> > 
> > Try removing the local log file, and then run the unit tests to see if that fails. I believe the logger doesn't execute the code that creates the logfile in the case of the unit tests so its logging to a file that doesn't exist.
> 
> There is no log file committed since it's ignored. I think it might be with the logs folder not existing. I've added a .gitkeep to create the logs folder. Hopefully, that works 🙏 🙏 🙏

@aidanjgriffiths WOW! It passed the tests! If I squash my commits, the tests will need to be run again, right? It took so long...

File: speaker_verification/__init__.py
Comment (aidanjgriffiths) [inline speaker_verification/__init__.py]: Why is this commented out?

File: speaker_verification/__init__.py
Comment (aidanjgriffiths) [inline speaker_verification/__init__.py]: same as L29

File: speaker_verification/deep_speaker/audio.py
Comment (aidanjgriffiths) [inline speaker_verification/deep_speaker/audio.py]: Minor but maybe we could omit `logs-` from the logfile name since its contained within the `logs` dir and uses the `.log` extension.
```suggestion
    log_file=f'{file_path}/../logs/{today.strftime("%d-%m-%Y")}.log'
```

File: speaker_verification/model_evaluation.py
Comment (aidanjgriffiths) [inline speaker_verification/model_evaluation.py]: I'd remove `import os` on L4 since you are only using `os.path.dirname` and `os.path.abspath` which is already imported on L7. Could you refactor your usage of these imports to use the L7 imports?

File: speaker_verification/model_evaluation.py
Comment (aidanjgriffiths) [inline speaker_verification/model_evaluation.py]:
```suggestion
file_path = dirname(abspath(__file__))
```

File: speaker_verification/model_evaluation.py
Comment (aidanjgriffiths) [inline speaker_verification/model_evaluation.py]:
```suggestion

```

File: speaker_verification/sql_utils.py
Comment (aidanjgriffiths) [inline speaker_verification/sql_utils.py:4]: Same changes as `model_evaluation`

File: speaker_verification/sql_utils.py
Comment (aidanjgriffiths) [inline speaker_verification/sql_utils.py]: I feel like this logfile string should be optional for `SpeakerVerificationLogger` with `f'{file_path}/logs/logs-{today.strftime("%d-%m-%Y")}.log'` being the default. That way we wouldn't have to import datetime, and os.path everywhere we want to log to the daily logfile.

File: speaker_verification/sql_utils.py
Comment (aidanjgriffiths) [inline speaker_verification/sql_utils.py]: I think this should be logged as at least the WARNING level but to be consistant with the other logs I'd recommend just using ERROR.
```suggestion
        logger.error("Failed to read data from sqlite table", error)
```

**Review by aidanjgriffiths — COMMENTED** [review]
Hey @sometheasiekswx , I've had a better look at the PR today, I'm happy with the implementation you've provided but theres a few things like unnessesary imports and commented out code that you should address before we should merge. 

Great Job 😄

File: speaker_verification/__init__.py
Comment (sometheasiekswx) [reply speaker_verification/__init__.py]: Ahhh... I'm not sure, I think I was testing, but good catch!

File: speaker_verification/__init__.py
Comment (sometheasiekswx) [reply speaker_verification/__init__.py]: Fixed it :thumbs up:

File: speaker_verification/deep_speaker/audio.py
Comment (sometheasiekswx) [reply speaker_verification/deep_speaker/audio.py]: Sure!

File: speaker_verification/model_evaluation.py
Comment (sometheasiekswx) [reply speaker_verification/model_evaluation.py]: Good catch! Will do that.

File: speaker_verification/sql_utils.py
Comment (sometheasiekswx) [reply speaker_verification/sql_utils.py:4]: 👍

File: speaker_verification/sql_utils.py
Comment (sometheasiekswx) [reply speaker_verification/sql_utils.py]: Good idea!

Comment (sometheasiekswx) [reply speaker_verification/sql_utils.py]: Looks much cleaner now!

File: speaker_verification/sql_utils.py
Comment (sometheasiekswx) [reply speaker_verification/sql_utils.py]: Yep you're right!

**Review by aidanjgriffiths — APPROVED** [review]
Nice @sometheasiekswx, the logger is working on my local machine and I'm happy with these changes you've made. Up to you whether you want to wait for @nnfunny to review or not before merging.

Thanks

**Pull request merged**
